File: 20240311_Monday/20240311.go
// Leetcode: 791. Custom Sort String.
// https://leetcode.com/problems/custom-sort-string/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 2 ms, faster than 29.69% of Go online submissions for Custom Sort
// String.
// Memory Usage: 2 MB, less than 86.98% of Go online submissions for Custom
// Sort String.
// 2024.03.11 Daily Challenge.

package main

import "sort"

func customSortString(order string, s string) string {
	// 1. Create sorting map.
	// 2. Split string, by chars in order and not.
	// 3. Sort.
	// 4. Concatinate.
	h := make(map[byte]int)
	ord := []byte(order)
	for i, v := range ord {
		h[byte(v)] = i
	}
	suffix := make([]byte, 0, len(s))
	prefix := make([]byte, 0, len(s))
	for _, v := range s {
		v := byte(v) // Shadowing.
		if _, ok := h[v]; ok {
			suffix = append(suffix, byte(v))
		} else {
			prefix = append(prefix, byte(v))
		}
	}
	sort.Slice(suffix, func(i, j int) bool {
		return h[suffix[i]] < h[suffix[j]]
	})
	return string(prefix) + string(suffix)
}

File: 20240311_Monday/20240311.rb
# Leetcode: 791. Custom Sort String.
# https://leetcode.com/explore/challenge/card/july-leetcoding-challenge-2021/609/week-2-july-8th-july-14th/3813/
# = = = = = = = = = = = = = 
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = =
# Runtime: 77 ms, faster than 5.09% of Ruby online submissions for Custom Sort
# String.
# Memory Usage: 211.4 MB, less than 77.97% of Ruby online submissions for
# Custom Sort String.
# 2024.03.11 Daily Challenge.

# @param {String} order
# @param {String} str
# @return {String}
def custom_sort_string(order, str)
  weight = Array.new(26)
  # Fill first part of weight by indices.
  (0...order.size).each do |i|
    code = order[i].ord - ?a.ord
    weight[code] = i
  end
  # Fill last part of weights by the counter startign just after the last order index.
  counter = order.size
  (0...26).each do |i|
    if weight[i].nil?
      weight[i] = counter
      counter += 1
    end
  end
  str.chars.sort_by{|x| weight[x.ord - ?a.ord]}.join
end
